pallets/xcm-core-buyer/src/lib.rs

```diff
@@ -382,23 +595,55 @@ pub mod pallet {
         fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
             let mut weight = Weight::zero();
 
-            // 1 write in on_finalize
-            weight += T::DbWeight::get().writes(1);
+            weight += T::DbWeight::get().writes((2 * T::MaxInFlightOrders::get() + 3) as u64);
+            weight += T::DbWeight::get().reads(3);
 
             weight
         }
 
-        fn on_finalize(_: BlockNumberFor<T>) {
-            // We clear this storage item because we only need it to prevent collators from buying
-            // more than one core for the same parathread in the same relay block
-            // TODO: with async backing, it is possible that two tanssi blocks are included in the
-            // same relay block, so this kill is not correct, should only kill the storage if the
-            // relay block number has changed
-            // TODO: this allows collators to send N consecutive messages to buy 1 core for the same
-            // parathread, as long as the parathread block is not included in pallet_author_noting.
-            // So a malicious collator could drain the parathread tank account by buying cores on
-            // every tanssi block but not actually producing any block.
-            InFlightOrders::<T>::kill();
+        fn on_finalize(n: BlockNumberFor<T>) {
+            let mut ttl_queue = InFlightOrdersTtl::<T>::get();
+            let mut maybe_oldest_entry;
+            let mut oldest_block_number;
+            let mut query_id;
+
+            let mut query_ids_to_remove = vec![];
+
+            loop {
+                maybe_oldest_entry = ttl_queue.first().copied();
+                (oldest_block_number, query_id) = if let Some(oldest_entry) = maybe_oldest_entry {
+                    oldest_entry
+                } else {
+                    break;
+                };
+
+                // No entry at current block number found
```

Comment:
There is one thing I am thinking. Instead of "automatically" doing this `on_finalize`,  can't we do it such that this is automatically cleaned when a collator tries to send the XCM message? If the current block number - block_number_last_message_sent > ttl then we insert the new one, otherwise we error.

Reply:
Hmm. That could work.

Reply:
I was worried about the storage bloat, but I think this is doable.

Reply:
@girazoki We do both. On response it is cleared. And if response is not received within ttl it is cleaned as well as fallback.

Reply:
but only when a collator injects a new one right? or is it automatically done?

Reply:
@girazoki Only when collator tries to inject new one. But it is not para id specific. It will try to clean up as many expired entry it could.

Reply:
humm I think this can be a bit dangreous if too many need to be cleaned. Is there a reason for this? I think we can only expire the one that the collator is trying to send

Reply:
@tmpolaczyk what do you think about this?